netty/src/main/java/io/grpc/transport/netty/Http2Negotiator.java

```diff
@@ -112,168 +97,252 @@ public static ChannelHandler serverTls(SSLEngine sslEngine) {
-    final ChannelInitializer<Channel> initializer = new ChannelInitializer<Channel>() {
-      @Override
-      public void initChannel(Channel ch) throws Exception {
-        ch.pipeline().addLast(upgrader);
-        ch.pipeline().addLast(completionHandler);
+    return new BufferUntilHttpUpgradeCompleteHandler(upgrader);
+  }
+
+  /**
+   * Returns a {@link ChannelHandler} that ensures that the {@code handler} is added to the
+   * pipeline writes to the {@link Channel} may happen immediately, even before it is active.
+   */
+  public static ChannelHandler plaintext(final ChannelHandler handler) {
+    return new BufferUntilChannelActiveHandler(handler);
+  }
+
+  /**
+   * Buffers all writes until either {@link #flushBufferedAndRemove(ChannelHandlerContext)} or
+   * {@link #failBufferedAndClose(ChannelHandlerContext)} is called. This handler allows us to
+   * write to a {@link Channel} before we are allowed to write to it officially i.e.
+   * before it's active or the TLS Handshake is complete.
+   */
+  private abstract static class BufferWritesHandler extends ChannelDuplexHandler {
+
+    private ChannelHandler[] handlers;
+    private Queue<ChannelWrite> bufferedWrites = new ArrayDeque<ChannelWrite>();
+    private boolean flushed;
+
+    /**
+     * @param handlers the ChannelHandlers are added to the pipeline on channelRegistered and
+     *                 before this handler.
+     */
+    BufferWritesHandler(ChannelHandler... handlers) {
+      this.handlers = handlers;
+    }
+
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+      /**
+       * This check is necessary as a channel may be registered with different eventloops during
+       * its lifetime and we only want to set up the pipeline once.
+       * We can't set up the pipeline by specifying a ChannelInitializer in the client bootstrap,
+       * because the initializer is not executed in the calling thread but on channelRegistered
+       * in the eventloop thread. So there is a race condition between threads immediately writing
+       * to the channel (from outside the eventloop) and the pipeline being set up i.e. a
+       * write task (writing to the HeadContext) may be created before the pipeline is set up.
+       */
+      if (handlers != null) {
+        ctx.pipeline().addFirst(handlers);
+        handlers = null;
       }
-    };
+      super.channelRegistered(ctx);
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+      failBufferedAndClose(ctx);
+      super.channelInactive(ctx);
+    }
 
-    return new Negotiation() {
-      @Override
-      public ChannelInitializer<Channel> initializer() {
-        return initializer;
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+        throws Exception {
+      /**
+       * This check is necessary as there is a race condition between threads writing to the
```

Comment:
Ah I see.  Yeah that's ugly :/.  Maybe something like this would be more clear:

```
This check handles a race condition between AbstractChannelHandlerContext.write
(in the calling thread) and the removal of this handler (in the event loop thread). The problem
occurs in this sequence:

1) The write method identifies the next ChannelHandlerContext (i.e. for this handler)
2) This handler removes itself from the pipeline
3) The write method delegates to the invoker to call the write method in the event loop thread. 

When this happens, we identify that this handler has been removed with "bufferedWrites == null".
```